ably/rest/auth.py

```diff
@@ -32,64 +30,84 @@ class Method:
+            
+            basic_key = "%s:%s" % (options.keyId, options.keyValue)
             basic_key = base64.b64encode(basic_key.encode('utf-8'))
             self.__basic_credentials = basic_key.decode('ascii')
-            return
+        elif not options.auth_token and not useToken:
+            raise AblyException(reason="Neither Token auth nor Basic auth can be setup. Provide a key, or a token, or the means to request a token.",
+                                status_code=400,
+                                code=40000)
+        
+        if useToken:
+            # Using token auth
+            self.__auth_method = Auth.Method.TOKEN
+
+            if options.auth_token:
+                self._token_details = TokenDetails(id=options.auth_token)
+            else:
+                self._token_details = None
+
+            if options.auth_callback:
+                log.debug("using token auth with auth_callback")
+            elif options.auth_url:
+                log.debug("using token auth with auth_url")
+            elif options.keyValue:
+                log.debug("using token auth with client-side signing")
+            elif options.auth_token:
+                log.debug("using token auth with supplied token only")
+            else:
+                # Not a hard error, but any operation requiring authentication
+                # will fail
+                log.debug("no authentication parameters supplied")
 
-        # Using token auth
-        self.__auth_method = Auth.Method.TOKEN
 
-        if options.auth_token:
-            self.__token_details = TokenDetails(id=options.auth_token)
-        else:
-            self.__token_details = None
-
-        if options.auth_callback:
-            log.debug("using token auth with auth_callback")
-        elif options.auth_url:
-            log.debug("using token auth with auth_url")
-        elif options.key_value:
-            log.debug("using token auth with client-side signing")
-        elif options.auth_token:
-            log.debug("using token auth with supplied token only")
-        else:
-            # Not a hard error, but any operation requiring authentication
-            # will fail
-            log.debug("no authentication parameters supplied")
+    def can_request_token(self):
+      if self.__auth_options.keyId and self.__auth_options.keyValue:
+          return True
+      if self.__auth_options.authUrl:
+          return True
+      if self.__auth_options.auth_callback:
+         return True
+      return False
+      
+    def shouldUseTokenAuth(self):
+        return self.__auth_options.useTokenAuth or self.__auth_options.clientId or self.__auth_options.authUrl or self.__auth_options.auth_callback 
+
 
     def authorise(self, force=False, **kwargs):
-        if self.__token_details:
-            if self.__token_details.expires > self._timestamp():
+        if self._token_details:
+            if self._token_details.expires > self._timestamp():
```

Comment:
Perhaps add a second or two buffer in there as it will expire by the time it reaches the server.